File: Week 1/cpp/5.arithmetic_operation.cpp
/*
 * WEEK 1 - C++ FUNDAMENTALS
 * Topic: Arithmetic Operations & Operator Precedence
 * File: 5.arithmetic_operation.cpp
 *
 * CONCEPT:
 * C++ shares Java's arithmetic semantics for the basic five operators
 * (+ - * / %). Integer division truncates toward zero. Mixing int and
 * double promotes to double automatically.
 *
 * KEY POINTS:
 *  - + - * / % are the standard arithmetic operators
 *  - int / int  -> truncated int (10/6 == 1)
 *  - int / 0    -> undefined behaviour (CRASH or SIGFPE), unlike Python's exception
 *  - double / int -> double (the int is promoted)
 *  - parentheses override default precedence; * / % bind tighter than + -
 *  - There is NO `**` operator — use `std::pow(base, exp)` from <cmath>
 *
 * SYNTAX:
 *   int q = 10 / 6;            // 1
 *   double r = 10.0 / 6.0;     // 1.6666...
 *   int rem = 10 % 6;          // 4
 *   double p = std::pow(2,10); // 1024.0
 *
 * DRY RUN:
 *  a=3, b=10
 *  c = b/(2*a)  -> 10/6 = 1   (int division)
 *  d = b/(2.0*a)-> 10/6.0 = 1.6666...
 *
 * COMPLEXITY: O(1) per op.
 */

#include <iostream>
#include <cmath>

int main() {
    int a = 3;
    int b = 10;

    int c = b / (2 * a);             // integer division, truncates
    std::cout << "b / (2*a) = " << c << '\n';                // 1

    double d = b / (2.0 * a);        // promoted to double
    std::cout << "b / (2.0*a) = " << d << '\n';              // 1.6666...

    std::cout << "b % (2*a)  = " << b % (2 * a) << '\n';      // 4

    std::cout << "10 + 20 * 30 = " << 10 + 20 * 30 << '\n';   // 610
    std::cout << "(10+20) * 30 = " << (10 + 20) * 30 << '\n'; // 900

    // Exponentiation requires the cmath function
    std::cout << "pow(2,10) = " << std::pow(2, 10) << '\n';   // 1024

    // Integer overflow demo (32-bit signed): the value silently wraps
    int big = 2'000'000'000;          // ' is a digit-separator (C++14)
    int wrapped = big + big;          // overflow -> undefined/implementation-defined
    std::cout << "2e9 + 2e9 (32-bit int) = " << wrapped << "  <- WRAPPED\n";

    return 0;
}

/*
 * NOTES:
 *  - Java and C++ both truncate int/int toward zero.
 *  - C++ has NO arbitrary-precision int. Use long long (64 bits) or libraries
 *    like Boost.Multiprecision for big numbers.
 *  - C++ `%` follows the sign of the dividend (matches Java; differs from Python).
 *  - C++ has no built-in exponent operator — `std::pow` returns double.
 *  - Division by zero on integers is undefined behaviour (UB), not an exception.
 */

File: Week 1/rust/5.arithmetic_operation.rs
/*
 * WEEK 1 - RUST FUNDAMENTALS
 * Topic: Arithmetic Operations & Operator Precedence
 * File: 5.arithmetic_operation.rs
 *
 * CONCEPT:
 * Rust uses the standard +, -, *, /, % operators. Unlike Python, `/`
 * follows the SAME RULE as Java/C++: integer-by-integer division
 * truncates toward zero. Mixing integer and float types is NOT implicit
 * — you must cast with `as`.
 *
 * KEY POINTS:
 *  - i32 / i32  -> i32 (truncated toward zero)
 *  - f64 / f64  -> f64
 *  - To mix:    `i as f64 / 2.0`
 *  - No `**` operator — use `i32::pow(base, exp)` or `f64::powi/f64::powf`
 *  - Division/modulo by zero with integers PANICS (Rust's term for crash)
 *  - For float: division by zero gives `inf` or `nan` (IEEE 754)
 *
 * SYNTAX:
 *   let c = b / (2 * a);            // i32 / i32 -> i32
 *   let d = b as f64 / (2 * a) as f64;
 *   let p = 2_i32.pow(10);          // 1024
 *   let q = 2_f64.powi(10);          // 1024.0
 *
 * DRY RUN:
 *  a=3, b=10
 *  c = b / (2*a) = 10 / 6 = 1  (truncated)
 *  d = b as f64 / (2.0 * 3.0) = 1.6666...
 */

fn main() {
    let a: i32 = 3;
    let b: i32 = 10;

    let c = b / (2 * a);                 // 1
    println!("b / (2*a) = {c}");

    let d = b as f64 / (2.0 * a as f64); // 1.6666...
    println!("b/(2.0*a) = {d}");

    println!("b % (2*a) = {}", b % (2 * a));   // 4

    println!("10 + 20 * 30  = {}", 10 + 20 * 30);
    println!("(10+20) * 30 = {}", (10 + 20) * 30);

    // Exponentiation — separate functions for integer/float
    println!("2_i32.pow(10)   = {}", 2_i32.pow(10));      // 1024
    println!("2_f64.powi(10) = {}", 2_f64.powi(10));     // 1024.0
    println!("2_f64.powf(0.5) = {}", 2_f64.powf(0.5));   // 1.4142...

    // Division-by-zero behaviour
    let zero = 0;
    // let bad = 1 / zero;             // would PANIC at runtime
    let inf = 1.0_f64 / 0.0;            // IEEE inf
    let nan = 0.0_f64 / 0.0;            // IEEE NaN
    println!("1.0 / 0.0 = {inf}, 0.0/0.0 = {nan}");

    // Overflow checking differences (debug vs release)
    let big: i32 = i32::MAX;
    let wrapped = big.wrapping_add(1);
    println!("i32::MAX.wrapping_add(1) = {wrapped}");

    let _ = zero;  // silence unused warning
}

/*
 * NOTES:
 *  - Rust matches Java in truncating integer division (Python uses // for this).
 *  - Implicit conversions: NONE. You must write `x as f64` or similar.
 *  - In debug builds, integer overflow PANICS; in release it wraps. Use
 *    `checked_*` (returns Option), `saturating_*`, or `wrapping_*` explicitly.
 *  - `%` follows the sign of the dividend (like Java), not the divisor.
 *  - For exact decimal arithmetic, use the `rust_decimal` crate (no std equivalent).
 */
